src/FastRouteRouter.php

```diff
@@ -267,44 +261,87 @@ public function generateUri($name, array $substitutions = [], array $options = [
+        $requiredParameters = [];
 
         // One route pattern can correspond to multiple routes if it has optional parts
-        foreach ($routes as $r) {
+        foreach ($routes as $parts) {
+            // Check if all parameters can be substituted
+            $requiredParameters = $this->hasRequiredParameters($parts, $substitutions);
+
+            // If not all parameters can be substituted, try the next route
+            if ($requiredParameters !== true) {
+                continue;
+            }
+
+            // Generate the path
             $path = '';
-            $paramIdx = 0;
-            foreach ($r as $part) {
-                // Fixed segment in the route
+            foreach ($parts as $part) {
                 if (is_string($part)) {
+                    // Append the string
                     $path .= $part;
                     continue;
                 }
 
-                // Placeholder in the route
-                if ($paramIdx === count($substitutions)) {
-                    throw new LogicException('Not enough parameters given');
-                }
-
-                if (! isset($substitutions[$part[0]])) {
-                    throw new Exception\InvalidArgumentException(
-                        'Optional segments with unsubstituted parameters cannot '
-                        . 'contain segments with substituted parameters when using FastRoute'
-                    );
+                // Check substitute value with regex
+                $regex = '~^' . str_replace('/', '\/', $part[1]) . '$~';
+                if (! preg_match($regex, $substitutions[$part[0]])) {
+                    throw new Exception\InvalidArgumentException(sprintf(
+                        'Parameter value for [%s] did not match the regex `%s`',
+                        $part[0],
+                        $part[1]
+                    ));
                 }
 
+                // Append the substituted value
                 $path .= $substitutions[$part[0]];
-                $paramIdx++;
             }
 
-            // If number of params in route matches with number of params given, use that route.
-            // Otherwise try to find a route that has more params
-            if ($paramIdx === count($substitutions)) {
-                return $path;
+            // Return generated path
+            return $path;
+        }
+
+        // No valid route was found; explain which minimal required parameters are needed
+        throw new Exception\InvalidArgumentException(sprintf(
+            'Expected parameter values for at least [%s], but received [%s]',
+            implode(',', $requiredParameters),
+            implode(',', array_keys($substitutions))
+        ));
+    }
+
+    /**
+     * Checks if all route parameters are available
+     *
+     * @param array $parts
+     * @param array $substitutions
+     *
+     * @return array|bool
+     */
+    private function hasRequiredParameters(array $parts, array $substitutions)
```

Comment:
Done. The function is now called `missingParameters()`  and always returns an array. If there are any parameters missing it will return those, otherwise it returns an empty array.